gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.util.concurrent;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multimap;
+import org.apache.commons.lang.StringUtils;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+import lombok.Synchronized;
+
+
+public abstract class TaskScheduler<K, T extends ScheduledTask<K>> implements Closeable {
+  private final Multimap<K, CancellableScheduledTask<K>> cancellableScheduledTaskMap = ArrayListMultimap.create();
+  private volatile boolean isStarted = false;
+
+  protected TaskScheduler() {
+  }
+
+  @Synchronized
+  final void start(Optional<String> name) {
+    if (!this.isStarted) {
+      this.startImpl(name);
+      this.isStarted = true;
+    }
+  }
+
+  abstract void startImpl(Optional<String> name);
+
+  public final void schedule(T task, long interval, TimeUnit timeUnit) {
+    Preconditions.checkArgument(isStarted, "TaskScheduler is not started");
+    CancellableScheduledTask<K> cancellable = scheduleImpl(task, interval, timeUnit);
+    this.cancellableScheduledTaskMap.put(task.getKey(), cancellable);
+  }
+
+  abstract CancellableScheduledTask<K> scheduleImpl(T task, long interval, TimeUnit timeUnit);
+
+  public final Iterable<CancellableScheduledTask<K>> getScheduledTasks() {
+    return ImmutableList.copyOf(this.cancellableScheduledTaskMap.values());
+  }
+
+  public final Iterable<CancellableScheduledTask<K>> getScheduledTasks(K key) {
+    return ImmutableList.copyOf(this.cancellableScheduledTaskMap.get(key));
+  }
+
+  public void cancel(CancellableScheduledTask<K> task) {
+    if (this.cancellableScheduledTaskMap.remove(task.getKey(), task)) {
+      task.cancel();
```

Comment:
How about `task.cancel()` can return a boolean and if it returns `true` we remove it from the collection and return `true`; otherwise, leave it in the collection and return `false`.